Add support for multiple module interfaces per `cc_library`

fmeum opened the pull request:
When multiple `module_interfaces` are specified on a single `cc_library`, the individual compilation actions form a DAG based on `import`s between these modules. Consider the following situation:

* `a.cppm` imports `b.cppm`, both of which are in the `module_interfaces` of a single `cc_library`.
* Building the target populates the action cache with an entry for `a.pcm` that stores `b.pcm` as a discovered input.
* Now edit `a.cppm` and `b.cppm` so that `b.cppm` imports `a.cppm` and `a.cppm` no longer imports `b.cppm`.
* Build again (optionally after a shutdown).

Before this commit, this resulted in an action cycle since during action cache checking, Bazel would reuse or look up the inputs discovered in the previous build, thus introducing an edge from `a.pcm` to `b.pcm`. Together with the newly discovered edge from `b.pcm` to `a.pcm`, this resulted in a cycle.

This is fixed by not requesting the previously discovered inputs (either retained in memory or in the action cache) if the mandatory inputs changed. In the case of C++20 modules, this is sufficient since the modmap file, which lists all transitive `.pcm` files required for compilation, is a mandatory input.

As part of this change, `MetadataDigestUtils.fromMetadata` had to be modified to always return a byte array of proper digest length, even if called with an empty map, to match the assumptions of the action cache.

fmeum committed:
Invalidate discovered inputs

fmeum committed:
TMP: Enable C++20 modules compilation test

fmeum committed:
Add cycle test

fmeum committed:
Update cc_integration_test.sh

lberki commented:
Disclaimer: so far, I have not participated in the implementation of C++20 modules at all, so this all is quite alien to me. You'd probably be better off getting an informed opinion from @pzembrod wrt. the C++ parts. I do feel competent enough to comment on the general mechanism, though.

I assume the "other related change" you mentioned over private chat is https://github.com/bazelbuild/bazel/pull/26859 ? Where is the associated to 

Ow, this is really thorny.  if I understand correctly,  it's not enough to persist the module map file (at least a single one) because there is no guarantee that the change that creates the cycle happens at the top level. E.g. if `a.ccpm` imports `b.cppm` imports `c.cppm`, you'd presumably include `a.cppm` in the action cache entry. But then what if the change is that `c.cppm` imports `b.cppm`? Then your file remains unchanged, so you are still susceptible to the same dependency cycle.

In addition to this, special-casing that one file is weird because then Bazel relies on explicit action by the rule author to stay correct. This isn't unheard of (e.g. persistent workers), but it's bad if even Skyframe required this kind of outside help.

The only possible approach I can see is to request the inputs from the action cache entry in the same order as they were requested by the original action and verify their up-to-dateness also gradually. That way, you avoid cycles without extra hints at the cost of some inefficiency (Skyframe restarts)

But the simplestest approach would be to disallow multiple `.cppm` files in a single `cc_library`, and this is where my ignorance about C++20 modules comes in. Would this be feasible?

fmeum commented:
> I assume the "other related change" you mentioned over private chat is https://github.com/bazelbuild/bazel/pull/26859 ? Where is the associated to

It's related in the sense that without that PR, the situation described in the current PR would always result in a cycle, even if there is no action cache entry: Both `a.pcm` and `b.pcm` are in the allowed derived inputs for the action producing the other. 

> if I understand correctly, it's not enough to persist the module map file (at least a single one) because there is no guarantee that the change that creates the cycle happens at the top level.

That's correct for the module interface file, but the file we digest here is the module map file, which contains the file names of _transitive_ `.pcm` file dependencies, not just the direct ones. I think that this should be good enough for invalidation since it describes the complete set of discovered inputs.

> The only possible approach I can see is to request the inputs from the action cache entry in the same order as they were requested by the original action and verify their up-to-dateness also gradually. That way, you avoid cycles without extra hints at the cost of some inefficiency (Skyframe restarts)

Yes, this should work. We could first request all mandatory inputs in a batch and then the discovered deps one by one, discarding the cache entry when one of them mismatches. This would require persisting the individual hashes of all discovered files, I think, where today we only persist their paths.

> But the simplestest approach would be to disallow multiple .cppm files in a single cc_library, and this is where my ignorance about C++20 modules comes in. Would this be feasible?

This is the current approach, `cc_library` currently fails when `module_interfaces` has more than one entry. This does make usage rather verbose compared to other build systems. But as the original PR demonstrated, those other build systems (in this case, ninja) also struggly with this kind of invalidation. @PikachuHyA Can you already assess how problematic this restriction is in practice?

PikachuHyA commented:
> But the simplestest approach would be to disallow multiple `.cppm` files in a single `cc_library`, and this is where my ignorance about C++20 modules comes in. Would this be feasible?

I don't think it is feasible for real-world C++20 Modules projects.

The PR https://github.com/bazelbuild/bazel/pull/22553 banned multiple module interface files in a single `cc_binary/cc_library`, but that would be too restrictive for modules-native projects. Real C++20 Modules projects commonly have multiple module interfaces or partitions in a single library target. For example, the modules-native version of [async_simple](https://github.com/PikachuHyA/async_simple) cannot be built with bazel after that PR  https://github.com/bazelbuild/bazel/pull/22553 merged. Forcing one module interface file (e.g. `foo.cppm`) per target would force awkward splits, duplicated BUILD files or unnatural organization. 

So a blanket ban on multiple module-interface files is not a practical default.
The restriction that each `cc_binary/cc_library`'s `module_interfaces` attribute may contain only a single module interface file should be removed.

ChuanqiXu9 commented:
> Allowing my ignorance to shine through, how are C++ modules different enough from `cc_library` rules that a 1:1 mapping is not feasible? Naively, one would think that "one module / one library" is a good philosophy.
> 
> IOW: Why do real C++20 Modules projects commonly have multiple module interfaces or partitions in a single library target?

Since in C++, a module can be (and generally should be) composed by multiple module interfaces. e.g, an `async_simple` module is consisted by module interfaces in  https://github.com/alibaba/async_simple/tree/CXX20Modules/async_simple_module like https://github.com/alibaba/async_simple/blob/CXX20Modules/async_simple_module/Common.cppm , https://github.com/alibaba/async_simple/blob/CXX20Modules/async_simple_module/Executor.cppm and https://github.com/alibaba/async_simple/blob/CXX20Modules/async_simple_module/Future.cppm

they implement different parts of the module.

fmeum committed:
Push new test

fmeum committed:
Switch to two-phase lookup

# Conflicts:
#	src/main/java/com/google/devtools/build/lib/actions/ActionCacheChecker.java

fmeum committed:
Fix tests and improve comments

fmeum committed:
Handle MissingArtifactValue in input lookup

fmeum committed:
Add Lukacs' patch

fmeum committed:
Ensure that empty digests are of the right size

fmeum committed:
Use patches

fmeum committed:
Update comment

fmeum committed:
Fix ActionCacheTest

lberki reviewed (commented):
Sweet! I'd be more comfortable if @tjgq also reviewed this file because he was involved with the action cache and the action execution machinery way more recently than I was, though. So for now, I won't add the "pls merge this" label.

fmeum commented on src/main/java/com/google/devtools/build/lib/actions/cache/MetadataDigestUtils.java:
```diff
@@ -28,6 +28,8 @@
 public final class MetadataDigestUtils {
   private MetadataDigestUtils() {}
 
+  private static final byte[] EMPTY_DIGEST = new Fingerprint().digestAndReset();
```
Added an explanation to the PR description

tjgq commented on src/test/shell/bazel/cc_integration_test.sh:
```diff
@@ -2293,6 +2294,152 @@ EOF
+    fi
+  fi
+
+  if [[ "$(uname -s)" == "Darwin" ]]; then
```
This can be just `if is_darwin`. Likewise for the other test case.

fmeum committed:
Address comments

fmeum committed:
Wrap SourceArtifactException

fmeum commented:
@lberki I resolved the conflict.

fmeum commented:
@lberki I botched the conflict resolution, but it should be good now.

fmeum commented:
> What alternatives do you have under "everything else"? (I can't think of any other than maybe you set idea but that would probably require tapping into some other deduplication mechanism that's already there so as not to waste CPU time)

Just the set idea, nothing else. I hope the refactoring turns out to be manageable.

rdong8 commented:
Are there any plans to bring this back? It's quite important for this to work for modules to be really usable.
